Python build broken on MacOS workers

https://source.cloud.google.com/results/invocations/bdff7795-ad08-4dcc-8fad-7b18704c3146/targets/github%2Fgrpc%2Faggregate_tests/tests

Failing 100% of builds starting from https://source.cloud.google.com/results/invocations/1587abd7-6d54-46d3-b35d-86cd38ac3690/targets/github%2Fgrpc%2Faggregate_tests/tests

```
 We could not diagnose your build failure. If you are unable to proceed, please file an issue at http://www.github.com/grpc/grpc with `[Python install]` in the title; please attach the whole log (including everything that may have appeared above the Python backtrace).
    Traceback (most recent call last):
      File "/Volumes/BuildData/tmpfs/tmp/pip-req-build-__RJ3e/src/python/grpcio/commands.py", line 272, in build_extensions
        build_ext.build_ext.build_extensions(self)
      File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_gevent/py27_gevent/lib/python2.7/site-packages/Cython/Distutils/old_build_ext.py", line 194, in build_extensions
        self.build_extension(ext)
      File "/Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_gevent/py27_gevent/lib/python2.7/site-packages/setuptools/command/build_ext.py", line 199, in build_extension
        _build_ext.build_extension(self, ext)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/command/build_ext.py", line 499, in build_extension
        depends=ext.depends)
      File "/Volumes/BuildData/tmpfs/tmp/pip-req-build-__RJ3e/src/python/grpcio/_parallel_compile_patch.py", line 43, in _parallel_compile
        output_dir, macros, include_dirs, sources, depends, extra_postargs)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/ccompiler.py", line 329, in _setup_compile
        raise TypeError, "'output_dir' must be a string or None"
    TypeError: 'output_dir' must be a string or None
```

python artifact build for python 3.4 and 3.5 is also broken:
https://source.cloud.google.com/results/invocations/58e48a4b-f844-423d-b4a3-f41f85736511/targets/github%2Fgrpc/tests;passed=true;group=tasks;test=build_artifact.python_macos_x64_python3.4;row=7

```
/usr/bin/clang -fno-strict-aliasing -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -arch i386 -arch x86_64 -g -DOPENSSL_NO_ASM=1 -D_WIN32_WINNT=1536 -DGPR_BACKWARDS_COMPATIBILITY_MODE=1 -DHAVE_CONFIG_H=1 -DGRPC_ENABLE_FORK_SUPPORT=1 -DPyMODINIT_FUNC=extern "C" __attribute__((visibility ("default"))) PyObject* -DGRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK=1 -Isrc/python/grpcio -Iinclude -I. -Ithird_party/boringssl/include -Ithird_party/zlib -Ithird_party/nanopb -Ithird_party/cares -Ithird_party/cares/cares -Ithird_party/cares/config_darwin -Ithird_party/address_sorting/include -I/Library/Frameworks/Python.framework/Versions/3.4/include/python3.4m -c src/python/grpcio/grpc/_cython/cygrpc.cpp -o python_build/temp.macosx-10.7-intel-3.4/src/python/grpcio/grpc/_cython/cygrpc.o -std=c++11 -fvisibility=hidden -fno-wrapv -fno-exceptions -DPB_FIELD_32BIT -pthread
In file included from src/python/grpcio/grpc/_cython/cygrpc.cpp:731:
In file included from ./src/core/lib/iomgr/resolve_address.h:40:
In file included from ./src/core/lib/iomgr/pollset_set.h:24:
In file included from ./src/core/lib/iomgr/pollset.h:27:
In file included from ./src/core/lib/iomgr/exec_ctx.h:31:
./src/core/lib/gprpp/memory.h:47:25: error: no member named 'forward' in namespace 'std'
  return new (p) T(std::forward<Args>(args)...);
                   ~~~~~^
./src/core/lib/gprpp/memory.h:47:33: error: 'Args' does not refer to a value
  return new (p) T(std::forward<Args>(args)...);
                                ^
./src/core/lib/gprpp/memory.h:44:35: note: declared here
template <typename T, typename... Args>
                                  ^
./src/core/lib/gprpp/memory.h:65:24: error: no type named 'unique_ptr' in namespace 'std'
using UniquePtr = std::unique_ptr<T, Deleter>;
                  ~~~~~^
./src/core/lib/gprpp/memory.h:65:34: error: expected ';' after alias declaration
using UniquePtr = std::unique_ptr<T, Deleter>;
                                 ^
                                 ;
./src/core/lib/gprpp/memory.h:68:8: error: unknown type name 'UniquePtr'
inline UniquePtr<T> MakeUnique(Args&&... args) {
       ^
./src/core/lib/gprpp/memory.h:68:17: error: expected unqualified-id
inline UniquePtr<T> MakeUnique(Args&&... args) {
                ^
6 errors generated.
```

--- a/src/python/grpcio/_parallel_compile_patch.py
+++ b/src/python/grpcio/_parallel_compile_patch.py
@@ -19,7 +19,6 @@ Enabling parallel build helps a lot.
 
 import distutils.ccompiler
 import os
-import six
 
 try:
     BUILD_EXT_COMPILER_JOBS = int(
@@ -38,12 +37,10 @@ def _parallel_compile(self,
                       extra_preargs=None,
                       extra_postargs=None,
                       depends=None):
-    if isinstance(output_dir, six.text_type):
-        output_dir = str(output_dir)
     # setup the same way as distutils.ccompiler.CCompiler
     # https://github.com/python/cpython/blob/31368a4f0e531c19affe2a1becd25fc316bc7501/Lib/distutils/ccompiler.py#L564
     macros, objects, extra_postargs, pp_opts, build = self._setup_compile(
-        output_dir, macros, include_dirs, sources, depends, extra_postargs)
+        str(output_dir), macros, include_dirs, sources, depends, extra_postargs)
     cc_args = self._get_cc_args(pp_opts, debug, extra_preargs)
 
     def _compile_single_file(obj):

--- a/src/python/grpcio/_parallel_compile_patch.py
+++ b/src/python/grpcio/_parallel_compile_patch.py
@@ -19,6 +19,7 @@ Enabling parallel build helps a lot.
 
 import distutils.ccompiler
 import os
+import six
 
 try:
     BUILD_EXT_COMPILER_JOBS = int(
@@ -37,6 +38,8 @@ def _parallel_compile(self,
                       extra_preargs=None,
                       extra_postargs=None,
                       depends=None):
+    if isinstance(output_dir, six.text_type):
+        output_dir = str(output_dir)
     # setup the same way as distutils.ccompiler.CCompiler
     # https://github.com/python/cpython/blob/31368a4f0e531c19affe2a1becd25fc316bc7501/Lib/distutils/ccompiler.py#L564
     macros, objects, extra_postargs, pp_opts, build = self._setup_compile(

